File: 53/js/53.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
    let n = nums.length;
    if (n === 1) {
        return nums[0];
    }
    
    let curS = nums[0];
    let golS = nums[0];

    for (let i = 1; i < n; i++){
        let num = nums[i];
        if (curS + num < num) {
            curS = num;
        }else{
            curS += num;
        }

        if (curS > golS) {
            golS = curS;
        }
    }

    return golS;
};

File: 53/go/53.go
// DP

// contiguous
// at every step i:
// if nums[i] > nums[i] + curS -> curS = nums[i] , else curS += nums[i]
// then:
// if golS < curS -> golS = curS

func maxSubArray(nums []int) int {
    // check 
    n := len(nums)
    if n == 1 {
        return nums[0]
    }

    curS := nums[0]
    golS := nums[0]

    for i := 1; i < n; i++ {
        num := nums[i]
        if num > num + curS {
            curS = num
        }else{
            curS += num
        }

        if golS < curS {
            golS = curS
        }
    }
    return golS
}